admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,444 @@
+  auto R_ref = admittance_state.ref_trans_base_ft.rotation();
+  auto R_desired = desired_trans_base_ft.rotation();
+  auto R = R_desired * R_ref.transpose();
+  auto angle_axis = Eigen::AngleAxisd(R);
+  X.block<3, 1>(3, 0) = angle_axis.angle() * angle_axis.axis();
+
+  // get admittance relative velocity
+  auto X_dot = Eigen::Matrix<double, 6, 1>(admittance_state.admittance_velocity.data());
+
+  // get external force
+  auto F_base = Eigen::Matrix<double, 6, 1>(admittance_state.wrench_base.data());
+
+  // zero out any forces in the control frame
+  Eigen::Matrix<double, 6, 1> F_control;
+  F_control.block<3, 1>(0, 0) = rot_base_control.transpose() * F_base.block<3, 1>(0, 0);
+  F_control.block<3, 1>(3, 0) = rot_base_control.transpose() * F_base.block<3, 1>(3, 0);
+  F_control = F_control.cwiseProduct(admittance_state.selected_axes);
+  F_base.block<3, 1>(0, 0) = rot_base_control * F_control.block<3, 1>(0, 0);
+  F_base.block<3, 1>(3, 0) = rot_base_control * F_control.block<3, 1>(3, 0);
+
+  // Compute admittance control law: F = M*a + D*v + S*x
+  Eigen::Matrix<double, 6, 1> X_ddot =
+    admittance_state.mass_inv.cwiseProduct(F_base - D * X_dot - K * X);
+  bool success = kinematics_->convert_cartesian_deltas_to_joint_deltas(
+    admittance_state.current_joint_pos, X_ddot, admittance_state.ft_sensor_frame,
+    admittance_state.joint_acc);
+
+  // integrate motion in joint space
+  admittance_state.joint_vel += admittance_state.joint_acc * dt;
+  admittance_state.joint_pos += admittance_state.joint_vel * dt;
+
+  // calculate admittance velocity corresponding to joint velocity
+  success &= kinematics_->convert_joint_deltas_to_cartesian_deltas(
+    admittance_state.current_joint_pos, admittance_state.joint_vel,
+    admittance_state.ft_sensor_frame, admittance_state.admittance_velocity);
+  success &= kinematics_->convert_joint_deltas_to_cartesian_deltas(
+    admittance_state.current_joint_pos, admittance_state.joint_acc,
+    admittance_state.ft_sensor_frame, admittance_state.admittance_acceleration);
+
+  return success;
+}
+
+void AdmittanceRule::process_wrench_measurements(
+  const geometry_msgs::msg::Wrench & measured_wrench,
+  const Eigen::Matrix<double, 3, 3> & sensor_world_rot,
+  const Eigen::Matrix<double, 3, 3> & cog_world_rot)
+{
+  Eigen::Matrix<double, 3, 2, Eigen::ColMajor> new_wrench;
+  new_wrench(0, 0) = measured_wrench.force.x;
+  new_wrench(1, 0) = measured_wrench.force.y;
+  new_wrench(2, 0) = measured_wrench.force.z;
+  new_wrench(0, 1) = measured_wrench.torque.x;
+  new_wrench(1, 1) = measured_wrench.torque.y;
+  new_wrench(2, 1) = measured_wrench.torque.z;
+
+  // transform to world frame
+  Eigen::Matrix<double, 3, 2> new_wrench_base = sensor_world_rot * new_wrench;
+
+  // apply gravity compensation
+  new_wrench_base(2, 0) -= ee_weight_[2];
+  new_wrench_base.block<3, 1>(0, 1) -= (cog_world_rot * cog_).cross(ee_weight_);
```

Comment:
We should do this inside a filter. Proposal is already in control_toolbox repository. They should be extended to work with Eigen too. Problem with this approach that only one gravity compensation can be used where in practice two filters are interesting, one for flange and other for tool.

Reply:
I think this could be a good idea, but do you think its needs to be done in this PR, especially since it is not merged yet?

Reply:
Issues is already opened (#424). We should do in the follow-up PR.